### DISCUSSION TITLE
Coordinate data returned from the live feed endpoint

### DISCUSSION THREAD

**Discussion by andersme** [Afischbacher/Nhl.Api#20] [Q&A]
Just curious if anyone has actually used this data to build anything. I'm having some issues with the X values and determining which end of the rink they are on and thus the distance from the goal. I know that the coordinates should generally be negative for left side and positive for right side. My issue is in trying to determine, given the period, which team should be at which end. Different rinks have different home sides so this makes this rather challenging. Any thoughts?

**Comment by Afischbacher** [comment]
Hey @andersme!

So there is the ability to see which side of the rink which team is on, take a look here when addressing which team is on which side

https://github.com/Afischbacher/Nhl.Api/blob/cb4d9b9d759ccf37f1e6c3dfb174029ac39e9277/Nhl.Api.Domain/Models/Game/LiveGameFeed.cs#L390

So you can then visualize the rink based on an x and y coordinate like this:

<img width="638" alt="image" src="https://user-images.githubusercontent.com/15982357/145835333-4f82b796-9bc3-4aab-8498-de916fa6dada.png">

So for example, when Ovechkin scores from his "office" he is usually around the area of the top of the circle on the left side.

Here is an example of the data the live game feed would provide.

```
{
    "event": "Goal",
    "eventCode": "PHI59",
    "eventTypeId": "GOAL",
    "description": "Alex Ovechkin (23) Slap Shot, assists: John Carlson (29), Evgeny Kuznetsov (15)",
    "secondaryType": "Slap Shot",
    "strength": {
        "code": "PPG",
        "name": "Power Play"
    },
    "gameWinningGoal": false,
    "emptyNet": false,
    "about": {
        "eventIdx": 18,
        "eventId": 59,
        "period": 1,
        "periodType": "REGULAR",
        "ordinalNum": "1st",
        "periodTime": "03:38",
        "periodTimeRemaining": "16:22",
        "dateTime": "2021-04-17T16:43:36Z",
        "goals": {
            "away": 1,
            "home": 0
        }
    },
    "coordinates": {
        "x": 51.0,
        "y": 22.0
    }
}
```
So his X coordinate is 51 and his Y coordinate is 22 on the ice.

Here is a video to add the visual representation of where he is on the ice when he scored https://www.nhl.com/islanders/video/ovechkin-scores-from-his-office/t-277521762/c-8097097

So if we were to put this on the map he would be roughly here
<img width="638" alt="image" src="https://user-images.githubusercontent.com/15982357/145838127-785e2c3b-05e2-4950-9694-151ed7cdaad6.png">

Hope this helps!

**Comment by andersme** [comment]
I see the property RinkSide in LiveGameFeedTeam, but I can't seem to find where that is used or how to access it. The classes that inherit from this base aren't used anywhere. Am I missing something?

**Reply by andersme** [reply]
Nevermind, I think I've figured out how to get to that property. Sadly, the API doesn't seem to always be correct. If you look at game 2010020005, the rink side appears to be incorrectly recorded. Have you experienced this ever or adjust for it?

**Reply by Afischbacher** [reply]
No I have never had to adjust for it, I am not sure if it is or isn't incorrectly recorded, My question to you is how you have determined that the sides of the rink each team was on for that game was recorded incorrect?

**Reply by andersme** [reply]
If we look at the feed for this game, the linescore says that the home team is on the left for periods 1 and 3. We would expect that the X coordinates for their shots would be in the positives. However, in looking at the events, we see:
```
{
    "players": [
        {
            "player": {
                "id": 8471208,
                "fullName": "Dustin Penner",
                "link": "/api/v1/people/8471208"
            },
            "playerType": "Shooter"
        },
        {
            "player": {
                "id": 8462147,
                "fullName": "Miikka Kiprusoff",
                "link": "/api/v1/people/8462147"
            },
            "playerType": "Goalie"
        }
    ],
    "result": {
        "event": "Shot",
        "eventCode": "EDM53",
        "eventTypeId": "SHOT",
        "description": "Dustin Penner Wrist Shot saved by Miikka Kiprusoff",
        "secondaryType": "Wrist Shot"
    },
    "about": {
        "eventIdx": 4,
        "eventId": 53,
        "period": 1,
        "periodType": "REGULAR",
        "ordinalNum": "1st",
        "periodTime": "00:30",
        "periodTimeRemaining": "19:30",
        "dateTime": "2010-10-08T02:16:58Z",
        "goals": {
            "away": 0,
            "home": 0
        }
    },
    "coordinates": {
        "x": -60.0,
        "y": -9.0
    },
    "team": {
        "id": 22,
        "name": "Edmonton Oilers",
        "link": "/api/v1/teams/22",
        "triCode": "EDM"
    }
}
```

As you can see this X value is negative. This is the same for all events for the whole game. This game might be a bit of an anomoly, because it is played at Rexall Place, which is kind of weird, because the TV cameras are on the same side as the team benches, which is opposite from all other arenas. But, this is not the only game where we see this. I can find more if you'd like. I feel like this is a fairly important issue as most libraries I've seen which consume this data, don't adjust for this properly. This can lead to erroneous results of data models which consume this data.

**Reply by Afischbacher** [reply]
Hey @andersme 

Let's investigate, shall we? 

So I managed to pull the entire game here, https://youtu.be/CLwAETa3PUw?t=1032 (quality is ok, but visible enough)

Here is the corresponding event from the live game feed

```
        "result" : {
          "event" : "Goal",
          "eventCode" : "EDM83",
          "eventTypeId" : "GOAL",
          "description" : "Gilbert Brule (1) Wrist Shot, assists: Ryan Whitney (1)",
          "secondaryType" : "Wrist Shot",
          "strength" : {
            "code" : "PPG",
            "name" : "Power Play"
          },
          "gameWinningGoal" : true,
          "emptyNet" : false
        },
        "about" : {
          "eventIdx" : 58,
          "eventId" : 83,
          "period" : 1,
          "periodType" : "REGULAR",
          "ordinalNum" : "1st",
          "periodTime" : "09:04",
          "periodTimeRemaining" : "10:56",
          "dateTime" : "2010-10-08T02:34:04Z",
          "goals" : {
            "away" : 0,
            "home" : 1
          }
        },
        "coordinates" : {
          "x" : -75.0,
          "y" : -23.0
        },
```

Looks like the coordinates are X: -75 and Y: -23

And if you place them on the map here, it looks like the shot happened around here

<img width="644" alt="image" src="https://user-images.githubusercontent.com/15982357/145892219-83f79165-585a-4c6a-9095-3a72a37583ac.png">

and if we look at the actual video

<img width="644" alt="image" src="https://user-images.githubusercontent.com/15982357/145892296-ec7dc6ef-bf8c-4c4e-a67e-49772aaf4c64.png">

I'd say **in this specific case** the coordinates look accurate? 

But I think it makes sense for the 1st period specifically for the events and the coordinates provided

<img width="644" alt="image" src="https://user-images.githubusercontent.com/15982357/145894168-210f80ae-54a2-4934-8a71-9f8e5b3d86d8.png">

Also here is the image of the event you provided
<img width="644" alt="image" src="https://user-images.githubusercontent.com/15982357/145894532-0b3603c0-9b1d-41eb-a4b7-6337c7603188.png">

and the coordinates are X: -60 and Y: -9 and here is the visual image of the shot being taken

<img width="644" alt="image" src="https://user-images.githubusercontent.com/15982357/145894774-7d64fce1-8542-4294-890d-f62356356999.png">

So overall, from these 2 pieces of data, the shot coordinates seem fairly accurate?

Am I missing something?

**Reply by andersme** [reply]
The coordinates themselves seem correct. However the RinkSide property from the linescore is wrong. According to the linescore, the home should be on the left. However, in your video, the away team is on the left.

For some context, what I'm trying to do is calculate distance to goal for each shot. In order to do so, I need to know which end the goal is at that the shooter is trying to shoot at. This is why I need to know the rink side, so that I can calculate it properly, otherwise, all the shots from the opposite side of the red line (which there are a decent amount) would be on the wrong side.

**Reply by Afischbacher** [reply]
Hey @andersme,

I will find a way to correctly display the correct rink side and add it to the next version of the NHL API, can you do me a favor and find me instances of live game feeds where the rink side is incorrect, I have one game, but can you provide me as many as you know of?

Thanks

**Reply by andersme** [reply]
I can do that for sure. I'll try to compile a list tonight.

**Reply by andersme** [reply]
[examples.csv](https://github.com/Afischbacher/Nhl.Api/files/7709158/examples.csv)

Here are some examples that my code detected. That's not nearly all of them, that's just up until 2013, but I had a bug that prevented me from getting more for now. I'll try to get the complete list tomorrow.

**Reply by andersme** [reply]
One thing to note, is that it appears that not all games have the rinkside property. Perhaps some games just didn't have it recorded. Maybe there's some other way to tell orientation other than using the linescore.

**Reply by andersme** [reply]
Here's the full list from 2010 to present:

[incorrectgames.csv](https://github.com/Afischbacher/Nhl.Api/files/7713098/incorrectgames.csv)

**Reply by Afischbacher** [reply]
Hey @andersme,

I am just writing unit tests around all of these cases, I am deciding whether I should preserve the original data from the NHL API or overwrite it, any thoughts here? I currently have an additional property, `CorrectedRinkSide` but there are cases where my logic won't be able to adjust for it, there are some games that have no play by play data, which omit's the ability for me to determine the coordinates of a play and determine which side which team in the rink they should be on. It covers 90%+ of the cases thus far, I will use your updated dataset to add to all of the test cases I currently have.

Thanks for the data and all your help, I will try and push out 1.7.0 this week with this fix.

Best Regards and hope this helps!

**Reply by andersme** [reply]
I currently have a work around in my code, and it's a bit of a hack. I simply look at the play-by-play events of each period and decide if they are mostly positive or negative X values for the home team. I then base my orientation on that. The seperate property for CorrectedRinkSide would be awesome! Thanks for all your help with this.

**Reply by Afischbacher** [reply]
I am basically doing the same, using a shot event and based on the home and away team and the x coordinate making the judgment on whether it's a home or away team and which side it should be on. I am using all the games you provided me as test cases, I will make a pull request and post it here to make sure my logic is not incorrect, but the unit tests should help cover and validate this for me.

Thanks again!
